### DISCUSSION TITLE
A mini tutorial?

### DISCUSSION THREAD

**Discussion by InfernalDread** [rhohndorf/Auto-Llama-cpp#4] [Q&A] [answered]
Hello,

First of all, thank you for creating this repo, I was hoping there would be people that share the desire in running auto gpt local! I am not a very tech savvy person, so I was wondering if anyone could make a short tutorial on getting everything up and running?

Thank you once again!

**Comment by thomasfifer** [comment, marked as answer]
Hi! I ended up trying this out recently and found a few issues but submitted a pull request for it. here's a quick tutorial of what I did on how to get it up and running with my changes. 

run these commands
>git clone https://github.com/rhohndorf/Auto-Llama-cpp.git
>cd Auto-Llama-cpp
>pip install -r requirements.txt

Download https://huggingface.co/eachadea/legacy-ggml-vicuna-13b-4bit/resolve/main/ggml-vicuna-13b-4bit.bin (8GB) and place it in your models folder (for testing I made one in the root directory of this project with another folder called 13B for the size but if you're trying out multiple llms you'll likely have a folder somewhere else for all of them)

Rename "env.template" file to ".env" changing any environment variables that you need (like the path to your recently downloaded model)

run this command to start
>python scripts/main.py

once you have things running and can see what it does, try changing ai_settings.yml and scripts/data/prompt.txt to change how the ai behaves

**Reply by InfernalDread** [reply]
Thank you! I will test this out as soon as I can and will report back if this worked for me as well. I should be able to test this today, so wish me luck!

**Reply by Alexander220799** [reply]
I get an Error Message after a couple of minutes. Created an Issue about that.

Maybe you can help?
https://github.com/rhohndorf/Auto-Llama-cpp/issues/10

**Comment by swisspitbull** [comment]
Hello, It is kinda working for me but i have this error at every step, nothing is working to fix it: 

ValueError: substring not found

12:36:58,856 AutoGPT INFO NEXT ACTION: : COMMAND = [36mError:[0m  ARGUMENTS = [36msubstring not found[0m
12:36:59,37 AutoGPT INFO SYSTEM: : Command Error: threw the following error: substring not found
12:41:48,176 AutoGPT INFO Error: 
: Traceback (most recent call last):
  File "G:\Auto-Llama-cpp\scripts\main.py", line 79, in print_assistant_thoughts
    assistant_reply_json = fix_and_parse_json(assistant_reply)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\Auto-Llama-cpp\scripts\json_parser.py", line 52, in fix_and_parse_json
    brace_index = json_str.index("{")
                  ^^^^^^^^^^^^^^^^^^^
ValueError: substring not found

12:41:48,623 AutoGPT INFO NEXT ACTION: : COMMAND = [36mError:[0m  ARGUMENTS = [36msubstring not found[0m
12:41:48,831 AutoGPT INFO SYSTEM: : Command Error: threw the following error: substring not found

**Reply by Alexander220799** [reply]
Exactly the Error I get. Created an Issue about this yesterday.

**Reply by swisspitbull** [reply]
Have you some output?? I got outputs someway yesterday using the main repository, but with this i get nothing.

**Reply by swisspitbull** [reply]
I get this error too: 
Warning: model not found. Using cl100k_base encoding

**Reply by thomasfifer** [reply]
what is the value of json_str at the point where you’re getting this error?

the warning is a nonissue. it’s just saying that this model isn’t part of tiktoken so we’re using a default encoding

**Reply by swisspitbull** [reply]
Nothing else in the log I can see, I am not a python dev, the error tracking is a little difficult for me but i can try.

**Reply by thomasfifer** [reply]
you could try running my latest pull request. it’s a one line change just turning on the debug output. if i remember right, it should print out the input before attempting to parse.

**Reply by swisspitbull** [reply]
`------------ JSON FIX ATTEMPT ---------------
Original JSON: ```json
{
        'thoughts': {
            'speak': 'Running the code...'
        },
        'command': {
            'name': 'evaluate_code',
            'args': {
                'code': code
            }
        }
    }
```
-----------
Fixed JSON:
----------- END OF FIX ATTEMPT ----------------
Failed to fix ai output, telling the AI.
Error: Invalid JSON
 # Max's GPT-3.5 Agent: Evaluate Code def evaluate_code(code): # Evaluate code by running it through a virtual machine and observing its output. max = { 'thoughts': { 'speak': 'Running the code...' }, 'command': { 'name': 'evaluate_code', 'args': { 'code': code } } } return max
MAX THOUGHTS:
REASONING:
CRITICISM:
json # Max's GPT-3.5 Agent: Evaluate Code
def evaluate_code(code):
    # Evaluate code by running it through a virtual machine and observing its output.
    max = {
        'thoughts': {
            'speak': 'Running the code...'
        },
        'command': {
            'name': 'evaluate_code',
            'args': {
                'code': code
            }
        }
    }
    return max
json loads error Expecting value: line 1 column 1 (char 0)
Warning: Failed to parse AI output, attempting to fix.
 If you see this warning frequently, it's likely that your prompt is confusing the AI. Try changing it up slightly.

llama_print_timings:        load time =  1847.93 ms
llama_print_timings:      sample time =     0.02 ms /     1 runs   (    0.02 ms per run)
llama_print_timings: prompt eval time = 24012.91 ms /   102 tokens (  235.42 ms per token)
llama_print_timings:        eval time =     0.00 ms /     1 runs   (    0.00 ms per run)
llama_print_timings:       total time = 24013.53 ms
------------ JSON FIX ATTEMPT ---------------
Original JSON: ```json
{
        'thoughts': {
            'speak': 'Running the code...'
        },
        'command': {
            'name': 'evaluate_code',
            'args': {
                'code': code
            }
        }
    }
```
-----------
Fixed JSON:
----------- END OF FIX ATTEMPT ----------------
Failed to fix ai output, telling the AI.
NEXT ACTION:  COMMAND = Error: ARGUMENTS = string indices must be integers, not 'str'`

**Reply by swisspitbull** [reply]
Maybe the issue??: "If you see this warning frequently, it's likely that your prompt is confusing the AI. Try changing it up slightly."

Edit: I try with the default prompts.

**Reply by swisspitbull** [reply]
This using default prompts:

`json loads error Expecting value: line 1 column 1 (char 0)
Error:
 Traceback (most recent call last): File "C:\AutoLLama\scripts\main.py", line 79, in print_assistant_thoughts assistant_reply_json = fix_and_parse_json(assistant_reply) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "C:\AutoLLama\scripts\json_parser.py", line 52, in fix_and_parse_json brace_index = json_str.index("{") ^^^^^^^^^^^^^^^^^^^ ValueError: substring not found
json GOALS:
1. Increase net worth.
2. Develop and manage multiple businesses autonomously.
3. Play to your strengths as an LLM and pursue simple strategies with no legal complications.

COMMANDS:
1. Google Search: "google", args: "input": "<search>"
5. Browse Website: "browse_website", args: "url": "<url>", "question": "<what_you_want_to_find_on_website>"
6. Start GPT Agent: "start_agent",  args: "name": "<name>", "task": "<short_task_desc>", "prompt": "<prompt>"
7. Message GPT Agent: "message_agent", args: "key": "<key>", "message": "<message>"
8. List GPT Agents: "list_agents", args: ""
9. Delete GPT Agent: "delete_agent", args: "key": "<key>"
10. Write to file: "write_to_file", args: "file": "<file>", "text": "<text>"
11. Read file: "read_file", args: "file": "<file>"
12. Append to file: "append_to_file", args: "file": "<file>", "text": "<text>"
13. Search Files: "search_files", args: "directory": "<directory>"
14. Evaluate Code: "evaluate_code", args: "code": "<full_code_string>"
15. Get Improved Code: "improve_code", args: "suggestions": "<list_of_suggestions>", "code": "<full_code_string>"
16. Write Tests: "write_tests", args: "code": "<full_code_string>", "focus": "<list_of_focus_areas>"
17. Execute Python File: "execute_python_file", args: "file": "<file>"
18. Task Complete (Shutdown): "task_complete", args: "reason": "<reason>"
19. Generate Image: "generate_image", args: "prompt": "<prompt>"
20. Do Nothing: "do_nothing", args: ""
}
json loads error Expecting value: line 1 column 1 (char 0)
NEXT ACTION:  COMMAND = Error: ARGUMENTS = substring not found`

**Comment by fenixlam** [comment]
I got another error after input the goal.
`
Traceback (most recent call last):
  File "D:\program\Auto-Llama-cpp\scripts\main.py", line 331, in <module>
    assistant_reply = chat.chat_with_ai(
  File "D:\program\Auto-Llama-cpp\scripts\chat.py", line 77, in chat_with_ai
    relevant_memory = permanent_memory.get_relevant(str(full_message_history[-5:]), 10)
  File "D:\program\Auto-Llama-cpp\scripts\memory\local.py", line 105, in get_relevant
    scores = np.dot(self.data.embeddings, embedding)
  File "<__array_function__ internals>", line 180, in dot
ValueError: shapes (0,8192) and (5120,) not aligned: 8192 (dim 1) != 5120 (dim 0)
`
In that area, that is local.get_relevent(text,k), I print the text variable and get an empty array [ ].
As I am just a noobie, it looks like that is not an error I can fix by myself... orz

**Comment by KenWuqianghao** [comment]
I'm getting this error, not sure whats the problem, any help appreciated
```
  File "/home/kenwu/anaconda3/envs/auto-llama/lib/python3.11/site-packages/llama_cpp/llama.py", line 150, in __init__
    assert self.ctx is not None
           ^^^^^^^^^^^^^^^^^^^^
AssertionError
```

**Reply by fenixlam** [reply]
> I'm getting this error, not sure whats the problem, any help appreciated
> 
> ```
>   File "/home/kenwu/anaconda3/envs/auto-llama/lib/python3.11/site-packages/llama_cpp/llama.py", line 150, in __init__
>     assert self.ctx is not None
>            ^^^^^^^^^^^^^^^^^^^^
> AssertionError
> ```
I have a work around. Goto the file scripts/memory/local.py
About line 107 you will see something like this:
```
embedding = get_ada_embedding(text)
scores = np.dot(self.data.embeddings, embedding)
top_k_indices = np.argsort(scores)[-k:][::-1]
return [self.data.texts[i] for i in top_k_indices]
```
The error is coming from text=="[]", so the work around is:
```
         if text!='[]':
            embedding = get_ada_embedding(text)
            scores = np.dot(self.data.embeddings, embedding)
            top_k_indices = np.argsort(scores)[-k:][::-1]
            return [self.data.texts[i] for i in top_k_indices]
        else:
            #scores=np.dot(self.data.embeddings,create_default_embeddings())
            scores=self.data.embeddings
            top_k_indices = np.argsort(scores)[-k:][::-1]
            return [self.data.texts[i] for i in top_k_indices]
```

**Comment by BainjAmAin-Bit** [comment]
Hi, I have a little problem that I can't solve. Thanks in advance for your help`Traceback (most recent call last):
  File "/Users/benjamin***********/Auto-Llama-cpp/scripts/main.py", line 3, in <module>
    import commands as cmd
  File "/Users/benjamin***********/Auto-Llama-cpp/scripts/commands.py", line 1, in <module>
    import browse
  File "/Users/benjamin***********Auto-Llama-cpp/scripts/browse.py", line 4, in <module>
    from llm_utils import create_chat_completion
  File "/Users/benjamin***********/Auto-Llama-cpp/scripts/llm_utils.py", line 5, in <module>
    llm = Llama(model_path=cfg.smart_llm_model, n_ctx=2048, embedding=True)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/llama_cpp/llama.py", line 184, in __init__
    raise ValueError(f"Model path does not exist: {model_path}")
ValueError: Model path does not exist: ./models/ggml-alpaca-7b-q4.bin
Exception ignored in: <function Llama.__del__ at 0x10686e020>
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/llama_cpp/llama.py", line 1309, in __del__
    if self.ctx is not None:
       ^^^^^^^^
AttributeError: 'Llama' object has no attribute 'ctx'`

I understood the access path was not good yet if

**Reply by jamey-rose** [reply]
To resolve this kind of issue I just used this model instead:
https://huggingface.co/TheBloke/Llama-2-13B-GGML/resolve/main/llama-2-13b.ggmlv3.q4_0.bin
